gateway/gateway.go

```diff
@@ -123,31 +125,38 @@ func NewGateway(
 
 // Reconfigure reconfigures the gateway.
 func (g *gateway) Reconfigure(ctx context.Context, deps resource.Dependencies, conf resource.Config) error {
-	cfg, err := resource.NativeConfig[*Config](conf)
-	if err != nil {
-		return err
-	}
-	// Determine if the pi is on bookworm or bullseye
-	osRelease, err := os.ReadFile("/etc/os-release")
-	if err != nil {
-		return fmt.Errorf("cannot determine os release: %w", err)
-	}
-	isBookworm := strings.Contains(string(osRelease), "bookworm")
-
-	g.bookworm = &isBookworm
-	g.rstPin = strconv.Itoa(*cfg.ResetPin)
+	g.mu.Lock()
+	defer g.mu.Unlock()
 
 	// If the gateway hardware was already started, stop gateway and the background worker.
 	// Make sure to always call stopGateway() before making any changes to the c config or
 	// errors will occur.
-	// Unexpected behavior will also occur if you call stopGateway() when the gateway hasn't been
+	// Unexpected behavior will also occur if stopGateway() is called when the gateway hasn't been
 	// started, so only call stopGateway if this module already started the gateway.
 	if g.started {
-		err := g.Close(ctx)
+		err := g.reset(ctx)
 		if err != nil {
 			return err
 		}
-		g.started = false
+	}
+
+	cfg, err := resource.NativeConfig[*Config](conf)
+	if err != nil {
+		return err
+	}
+	if len(deps) == 0 {
+		return errors.New("must add raspberry pi board as dependency")
+	}
+	var dep resource.Resource
+
+	// Assuming there's only one dep.
+	for _, val := range deps {
+		dep = val
+	}
```

Comment:
This is wrong if a user adds another dep that is also a board, it's non-deterministic because the assumption that there's only one dep is wrong. Either refuse to build without one dep or check all the dependencies to ensure that only one is a board and use that one. This, however, is confusing to troubleshoot and ripe for useability bugs if a user adds two board dependencies to their config.

Could you use the board?FromDependencies with an implicit dependency addition instead? It will break configs as it adds a required attribute. It wasn't in your scope doc either, so it's worth a call out in that document as well. We still have very low useability of this module, so changing those configs to the new version is priority number one after this version's release.

Also, please move this module code to viam-modules so that tickets can be tracked on Jira; it's already public, unfortunately, so there's no added benefit to having it in your GitHub org anymore.

Reply:
sure, this is already a breaking change because a user will need to add the board to the depends_on, so we might as well make it an implicit dependency with board attribute. Also want to point out that im doing the exact same thing in node.go to get the gateway dependency, so may be worth changing the gateway to be an implicit dependency before we have many users.